File: ios/src/ReactNative/RNIContextMenuButton/RNIContextMenuButtonManager.m
//
//  RNIContextMenuButtonManager.m
//  IosContextMenuExample
//
//  Created by Dominic Go on 10/28/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import <React/RCTViewManager.h>
 

@interface RCT_EXTERN_MODULE(RNIContextMenuButtonManager, RCTViewManager)

// MARK: - Export Props - Events
// -----------------------------

// TODO: WIP

RCT_EXPORT_VIEW_PROPERTY(onMenuWillShow  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuWillHide  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuWillCancel, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onMenuDidShow  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuDidHide  , RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMenuDidCancel, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onPressMenuItem, RCTBubblingEventBlock);

RCT_EXPORT_VIEW_PROPERTY(onRequestDeferredElement, RCTBubblingEventBlock);

// MARK: - Export Props - Values
// ----------------------------

RCT_EXPORT_VIEW_PROPERTY(menuConfig, NSDictionary);

RCT_EXPORT_VIEW_PROPERTY(enableContextMenu   , BOOL);
RCT_EXPORT_VIEW_PROPERTY(isMenuPrimaryAction , BOOL);
RCT_EXPORT_VIEW_PROPERTY(isContextMenuEnabled, BOOL);

// MARK: -  View Manager Commands
// ------------------------------

RCT_EXTERN_METHOD(dismissMenu:(nonnull NSNumber *)node);

@end

File: ios/src/ReactNative/RNIContextMenuButton/RNIContextMenuButtonManager.swift
//
//  RNIContextMenuButtonManager.swift
//  IosContextMenuExample
//
//  Created by Dominic Go on 10/28/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import UIKit;


@objc(RNIContextMenuButtonManager)
class RNIContextMenuButtonManager: RCTViewManager {
  
  override static func requiresMainQueueSetup() -> Bool {
    return true;
  };
  
  override func view() -> UIView! {
    if #available(iOS 14, *) {
      return RNIContextMenuButton(bridge: self.bridge);
      
    } else {
      return RCTView();
    };
  };
  
  @objc func dismissMenu(_ node: NSNumber) {
    DispatchQueue.main.async {
      guard #available(iOS 14, *),
            let view = self.bridge.uiManager.view(forReactTag: node),
            let contextMenuButton = view as? RNIContextMenuButton
      else { return };
      
      contextMenuButton.dismissMenu();
    };
  };
};
